bee-network/bee-autopeering/src/server.rs

```diff
@@ -0,0 +1,290 @@
+                                    peering_tx.send(packet).expect("channel send error: peering");
+                                }
+                                _ => panic!("invalid message type"),
+                            }
+                        }
+                        Err(e) => {
+                            log::error!("UDP socket read error; stopping incoming packet handler. Cause: {}", e);
+                            // TODO: intiate shutdown of the system
+                            break 'recv;
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Runnable for OutgoingPacketHandler {
+    const NAME: &'static str = "OutgoingPacketHandler";
+    const SHUTDOWN_PRIORITY: u8 = 3;
+
+    type ShutdownSignal = ShutdownRx;
+
+    async fn run(self, mut shutdown_rx: Self::ShutdownSignal) {
+        let OutgoingPacketHandler {
+            outgoing_socket,
+            mut outgoing_rx,
+            local,
+            bind_addr,
+        } = self;
+
+        'recv: loop {
+            tokio::select! {
+                _ = &mut shutdown_rx => {
+                    break;
+                }
+                o = outgoing_rx.recv() => {
+                    if let Some(packet) = o {
+                        let OutgoingPacket {
+                            msg_type,
+                            msg_bytes,
+                            peer_addr,
+                        } = packet;
+
+                        if peer_addr == bind_addr {
+                            log::warn!("Trying to send to own bind address: {}. Ignoring packet.", peer_addr);
+                            continue 'recv;
+                        }
+
+                        let marshalled_bytes = marshal(msg_type, &msg_bytes);
+
+                        let signature = local.sign(&marshalled_bytes);
+                        let packet = Packet::new(msg_type, &marshalled_bytes, &local.public_key(), signature);
+
+                        let bytes = packet.to_protobuf().expect("error encoding outgoing packet");
+
+                        if bytes.len() > MAX_PACKET_SIZE {
+                            log::warn!("Trying to send too many bytes to {}. Ignoring...", peer_addr);
+                            continue 'recv;
+                        }
+
+                        let n = outgoing_socket.send_to(&bytes, peer_addr).await.expect("socket send error");
+
+                        log::trace!("Sent {} bytes to {}.", n, peer_addr);
+                    } else {
+                        // All `outgoing_tx` message senders were dropped.
+                        break 'recv;
+                    }
+                }
+            }
+        }
+    }
+}
+
+pub(crate) fn marshal(msg_type: MessageType, msg_bytes: &[u8]) -> Vec<u8> {
+    let mut marshalled_bytes = vec![0u8; msg_bytes.len() + 1];
+    marshalled_bytes[0] = msg_type as u8;
```

Comment:
maybe `marshalled_bytes.split_at_mut(0)` could be better

Reply:
Can you describe in what way? Because I end up doing the same as before (except for the slightly different offset), but maybe I am just not seeing your idea (already late here). 

`split_at_mut` gives me two unique pointers into one array (in case I don't need overlapping slices), right? But I actually don't need that as I can do the copy one after the other. As much as I don't like how that was done in the Go impl, but I have to bite into the sour apple (German saying), and copy the data. I cannot see why the current code would be compiled to something less efficient :shrug:

Reply:
Closing this for now (added a comment as a reminder for Christian in case he wants to make a PR to optimize this they way he means to)